src/util/netif.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2024 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#include <config/bitcoin-config.h> // IWYU pragma: keep
+
+#include <util/netif.h>
+
+#include <logging.h>
+#include <netbase.h>
+#include <util/readwritefile.h>
+#include <util/sock.h>
+#include <util/strencodings.h>
+
+#ifdef HAVE_LINUX_ROUTE_H
+
+#include <linux/route.h>
+
+std::optional<CNetAddr> QueryDefaultGateway(Network network)
```

Comment:
The following patch uses a netlink socket to get the information from the kernel, that is supported on (at least) Linux and FreeBSD>=13.2:

<details>
<summary>[patch] get default gateway using a netlink socket</summary>

```diff
--- a/src/test/netbase_tests.cpp
+++ b/src/test/netbase_tests.cpp
@@ -1,25 +1,34 @@
 // Copyright (c) 2012-2022 The Bitcoin Core developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <compat/compat.h>
 #include <net_permissions.h>
 #include <netaddress.h>
 #include <netbase.h>
 #include <netgroup.h>
 #include <protocol.h>
 #include <serialize.h>
 #include <streams.h>
 #include <test/util/setup_common.h>
 #include <util/strencodings.h>
+#include <util/syserror.h>
 #include <util/translation.h>
 
 #include <string>
 
 #include <boost/test/unit_test.hpp>
 
+#ifdef __linux__
+#include <linux/rtnetlink.h>
+#elif defined(__FreeBSD__)
+#include <netlink/netlink.h>
+#include <netlink/netlink_route.h>
+#endif
+
 using namespace std::literals;
 
 BOOST_FIXTURE_TEST_SUITE(netbase_tests, BasicTestingSetup)
 
 static CNetAddr ResolveIP(const std::string& ip)
 {
@@ -610,7 +619,100 @@ BOOST_AUTO_TEST_CASE(isbadport)
             ++total_bad_ports;
         }
     }
     BOOST_CHECK_EQUAL(total_bad_ports, 80);
 }
 
+std::optional<CNetAddr> QueryDefaultGateway(Network network)
+{
+    Assume(network == NET_IPV4 || network == NET_IPV6);
+
+    // Create a netlink socket.
+
+    const int s{socket(AF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE)};
+    if (s < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "socket(AF_NETLINK): %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+    Sock sock{static_cast<SOCKET>(s)};
+
+    // Send request.
+
+    struct {
+        nlmsghdr hdr; ///< Request header.
+        rtmsg data; ///< Request data, a "route message".
+        nlattr dst_hdr; ///< One attribute, conveying the route destination address.
+        char dst_data[16]; ///< Route destination address. To query the default route we use 0.0.0.0/0 or [::]/0. For IPv4 the first 4 bytes are used.
+    } request;
+
+    // Whether to use the first 4 or 16 bytes from request.attr_dst_data.
+    const size_t dst_data_len = network == NET_IPV4 ? 4 : 16;
+
+    memset(&request, 0x0, sizeof(request));
+
+    request.hdr.nlmsg_type = RTM_GETROUTE;
+    request.hdr.nlmsg_flags = NLM_F_REQUEST;
+#ifdef __linux__
+    request.hdr.nlmsg_flags |= NLM_F_DUMP;
+#endif
+    request.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(rtmsg) + sizeof(nlattr) + dst_data_len);
+    request.hdr.nlmsg_seq = 0; // Sequence number, used to match which reply is to which request. Irrelevant for us because we send just one request.
+    request.data.rtm_family = network == NET_IPV4 ? AF_INET : AF_INET6;
+    request.data.rtm_dst_len = 0; // Prefix length.
+#ifdef __FreeBSD__
+    request.data.rtm_flags = RTM_F_PREFIX;
+#endif
+    request.dst_hdr.nla_type = RTA_DST;
+    request.dst_hdr.nla_len = sizeof(nlattr) + dst_data_len;
+
+    if (sock.Send(&request, sizeof(request), 0) != sizeof(request)) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "send() to netlink socket: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+
+    // Receive response.
+
+    char response[4096];
+    ssize_t response_len;
+    do {
+        response_len = sock.Recv(response, sizeof(response), 0);
+    } while (response_len < 0 && (errno == EINTR || errno == EAGAIN));
+    if (response_len < 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "recv() from netlink socket: %s\n", SysErrorString(errno));
+        return std::nullopt;
+    }
+
+    for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, response_len); hdr = NLMSG_NEXT(hdr, response_len)) {
+        rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
+        int remaining_len = RTM_PAYLOAD(hdr);
+        // Iterate over the attributes.
+        for (rtattr* attr = RTM_RTA(r); RTA_OK(attr, remaining_len); attr = RTA_NEXT(attr, remaining_len)) {
+            if (attr->rta_type == RTA_GATEWAY) {
+                if (network == NET_IPV4) {
+                    Assume(sizeof(in_addr) == RTA_PAYLOAD(attr));
+                    return CNetAddr{in_addr{.s_addr = *static_cast<decltype(in_addr::s_addr)*>(RTA_DATA(attr))}};
+                } else {
+                    Assume(sizeof(in6_addr) == RTA_PAYLOAD(attr));
+                    in6_addr gw;
+                    std::memcpy(&gw, RTA_DATA(attr), sizeof(gw));
+                    return CNetAddr{gw};
+                }
+            }
+        }
+    }
+
+    return std::nullopt;
+}
+
+BOOST_AUTO_TEST_CASE(netlink)
+{
+    for (const auto net : {NET_IPV4, NET_IPV6}) {
+        const auto gw{QueryDefaultGateway(net)};
+        if (gw.has_value()) {
+            printf("Default %s gateway: %s\n", GetNetworkName(net).c_str(), gw->ToStringAddr().c_str());
+        } else {
+            printf("No %s default gateway.\n", GetNetworkName(net).c_str());
+        }
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
```
</details>

I find the netlink interface a somewhat difficult to grasp.

https://man7.org/linux/man-pages/man7/netlink.7.html
https://man7.org/linux/man-pages/man7/rtnetlink.7.html
https://man7.org/linux/man-pages/man3/rtnetlink.3.html
https://stackoverflow.com/questions/11788326/extract-current-route-from-netlink-message-code-attached
https://www.rfc-editor.org/rfc/rfc3549

Reply:
Huh interesting. i didn't know netlink worked for multiple operating systems, that's much better.

Reply:
Is it possible to get the `scope_id` for IPv6 addresses? At least my router gives me an link-scope address.

Reply:
Here is a standalone program to get the default gateway using a netlink socket:

<details>
<summary>netlink_get_default_route.cc</summary>

```cc
#include <arpa/inet.h>
#include <assert.h>
#include <errno.h>
#include <net/if.h>
#include <netinet/in.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>

#ifdef __linux__
#include <linux/rtnetlink.h>
#elif defined(__FreeBSD__)
#include <netlink/netlink.h>
#include <netlink/netlink_route.h>
#endif

void QueryDefaultGateway(sa_family_t family)
{
    // Create a netlink socket.

    int sock = socket(AF_NETLINK, SOCK_DGRAM, NETLINK_ROUTE);
    if (sock < 0) {
        perror("socket(AF_NETLINK)");
        return;
    }

    // Send request.

    struct {
        nlmsghdr hdr; ///< Request header.
        rtmsg data; ///< Request data, a "route message".
        nlattr dst_hdr; ///< One attribute, conveying the route destination address.
        char dst_data[16]; ///< Route destination address. To query the default route we use 0.0.0.0/0 or [::]/0. For IPv4 the first 4 bytes are used.
    } request{};

    // Whether to use the first 4 or 16 bytes from request.attr_dst_data.
    const size_t dst_data_len = family == AF_INET ? 4 : 16;

    request.hdr.nlmsg_type = RTM_GETROUTE;
    request.hdr.nlmsg_flags = NLM_F_REQUEST;
#ifdef __linux__
    // XXX some strange behavior:
    // Linux IPv4 - this must be present, otherwise no gateway is found
    // Linux IPv6 - this must be present, otherwise no gateway is found
    // FreeBSD IPv4 - does not matter, the gateway is found with or without this
    // FreeBSD IPv6 - this must be absent, otherwise no gateway is found
    request.hdr.nlmsg_flags |= NLM_F_DUMP;
#endif
    request.hdr.nlmsg_len = NLMSG_LENGTH(sizeof(rtmsg) + sizeof(nlattr) + dst_data_len);
    request.hdr.nlmsg_seq = 0; // Sequence number, used to match which reply is to which request. Irrelevant for us because we send just one request.
    request.data.rtm_family = family;
    request.data.rtm_dst_len = 0; // Prefix length.
    //request.data.rtm_table = RT_TABLE_MAIN;
    //request.data.rtm_protocol = RTPROT_STATIC;
    //request.data.rtm_scope = RT_SCOPE_UNIVERSE;
    //request.data.rtm_type = RTN_UNICAST;
#ifdef __FreeBSD__
    // XXX some strange behavior:
    // Linux IPv4 - this must be absent, otherwise no gateway is found
    // Linux IPv6 - this must be absent, otherwise no gateway is found
    // FreeBSD IPv4 - does not matter, the gateway is found with or without this
    // FreeBSD IPv6 - this must be present, otherwise no gateway is found
    request.data.rtm_flags = RTM_F_PREFIX;
#endif
    request.dst_hdr.nla_type = RTA_DST;
    request.dst_hdr.nla_len = sizeof(nlattr) + dst_data_len;

    if (send(sock, &request, request.hdr.nlmsg_len, 0) != request.hdr.nlmsg_len) {
        fprintf(stderr, "send() failed to send %u bytes\n", request.hdr.nlmsg_len);
        close(sock);
        return;
    }

    // Receive response.

    char response[4096];
    ssize_t response_len;
    do {
        response_len = recv(sock, response, sizeof(response), 0);
    } while (response_len < 0 && (errno == EINTR || errno == EAGAIN));
    if (response_len < 0) {
        fprintf(stderr, "recv(): %s\n", strerror(errno));
        close(sock);
        return;
    }

    for (nlmsghdr *hdr = (nlmsghdr*)response; NLMSG_OK(hdr, response_len); hdr = NLMSG_NEXT(hdr, response_len)) {
        rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
        int remaining_len = RTM_PAYLOAD(hdr);

        // Iterate over the attributes.
        rtattr* rta_gateway = nullptr;
        int scope_id = 0;
        for (rtattr *attr = RTM_RTA(r); RTA_OK(attr, remaining_len); attr = RTA_NEXT(attr, remaining_len)) {
            if (attr->rta_type == RTA_GATEWAY) {
                rta_gateway = attr;
            } else if (attr->rta_type == RTA_OIF) {
                assert(sizeof(int) == RTA_PAYLOAD(attr));
                memcpy(&scope_id, RTA_DATA(attr), sizeof(scope_id));
            }
        }

        // Found gateway?
        if (rta_gateway != nullptr) {
            char buf[256];
            printf("%s gateway: %s",
                   family == AF_INET ? "IPv4" : "IPv6",
                   inet_ntop(r->rtm_family, RTA_DATA(rta_gateway), buf, sizeof(buf)));
            if (family == AF_INET6) {
                printf(", scope id: %d\n", scope_id);
            } else {
                printf("\n");
            }
        }
    }

    close(sock);
}

int main(int argc, char** argv)
{
    QueryDefaultGateway(AF_INET);
    QueryDefaultGateway(AF_INET6);
    return 0;
}
```
</details>

Ideally the `#ifdef __linux__` / `#ifdef __FreeBSD__` parts should not be needed. I am not sure if this is due to the above program doing something wrong or is due to a difference in Linux vs FreeBSD implementations. @AlexanderChernikov, @markjdb, any ideas?